### ISSUE TITLE
ADO.NET provider: VirtuosoDataReader.Read() hangs

### ISSUE THREAD

**Issue by drassokhin** [openlink/virtuoso-opensource#419]
VirtuosoDataReader.Read() hangs after executing a query while iterating over records in the result set. The behavior occurs randomly and is hard to reproduce. It happens more often when iterating over results of complex SPARQL queries that take a minute or more to execute. Here is a typical call stack (PollRead never returns):

OpenLink.Data.Virtuoso.dll!OpenLink.Data.Virtuoso.TcpSession.PollRead(int microSeconds = -1) + 0x8b bytes  
    OpenLink.Data.Virtuoso.dll!OpenLink.Data.Virtuoso.Future.GetResult(OpenLink.Data.Virtuoso.ISession session = {OpenLink.Data.Virtuoso.TcpSession}, OpenLink.Data.Virtuoso.FutureList futures = {OpenLink.Data.Virtuoso.FutureList}, bool remove = true) + 0x49e bytes  
    OpenLink.Data.Virtuoso.dll!OpenLink.Data.Virtuoso.Future.GetNextResult(OpenLink.Data.Virtuoso.ISession session = {OpenLink.Data.Virtuoso.TcpSession}, OpenLink.Data.Virtuoso.FutureList futures = {OpenLink.Data.Virtuoso.FutureList}) + 0x3f bytes 
    OpenLink.Data.Virtuoso.dll!OpenLink.Data.Virtuoso.ManagedCommand.ProcessResult(bool needEvl = true) + 0xb3 bytes  
    OpenLink.Data.Virtuoso.dll!OpenLink.Data.Virtuoso.ManagedCommand.Fetch() + 0x24d bytes  
    OpenLink.Data.Virtuoso.dll!OpenLink.Data.Virtuoso.VirtuosoDataReader.Read() + 0x75 bytes

**Comment by HughWilliams** [comment]
Is it the ADO.Net Provider hanging or the Virtuoso server i.e. can you still access the Virtuoso server via isql or the command line tools ?

If so then   can you please provide the output of running the "status()" command from the Virtuoso isql command line tool ?

**Comment by drassokhin** [comment]
Hugh --

It's the ADO.NET provider hanging (waiting for the Socket.Poll call to return):
`
    System.dll!System.Net.Sockets.Socket.Poll(int microSeconds, System.Net.Sockets.SelectMode mode) + 0x19a bytes   
    OpenLink.Data.Virtuoso.dll!OpenLink.Data.Virtuoso.TcpSession.PollRead(int microSeconds = -1) + 0x8b bytes   
`
I ran the same test against the same Virtuoso instance several times with the same SPARQL query (even restarted the Virtuoso service once between the test runs), and the provider never finished retrieving the complete recordset. It was getting stuck at PollRead after retrieving a relatively small but random number of records (51, 100, etc). It definitely has something to do with the timing. When I tested simple queries (even those that would return a large number of records), the client code usually ran to completion. Complicated queries that required more processing time on the server were much more likely to cause the client to hang. I experienced no problems when I ran the same queries via the SPARQL protocol (HTTP GET).

Here's the isql status() output produced when the ADO.NET client was hanging inside the PollRead method:

OpenLink Interactive SQL (Virtuoso), version 0.9849b.
Type HELP; for help and EXIT; to exit.
SQL> status();
Connected to OpenLink Virtuoso
Driver: 07.10.3211 OpenLink Virtuoso ODBC Driver
REPORT
VARCHAR

---

OpenLink Virtuoso  Server
Version 07.10.3211-threads for Win64 as of Feb 12 2015
Started on: 2015-05-25 13:59 GMT-4

Database Status:
  File size 0, 4410112 pages, 2109374 free.
  1360000 buffers, 274896 used, 2 dirty 0 wired down, repl age 0 0 w. io 0 w/crs
r.
  Disk Usage: 277733 reads avg 0 msec, 0% r 0% w last  0 s, 45310 writes flush
    266.7 MB/,
    996 read ahead, batch = 269.  Autocompact 0 in 0 out, 0% saved.
Gate:  7348 2nd in reads, 0 gate write waits, 0 in while read 0 busy scrap.
Log = virtuoso.trx, 3494 bytes
2300608 pages have been changed since last backup (in checkpoint state)
Current backup timestamp: 0x0000-0x00-0x00
Last backup date: unknown
Clients: 8 connects, max 3 concurrent
RPC: 584 calls, -4 pending, 19 max until now, 156 queued, 1 burst reads (0%), 32
7 second 0M large, 1046M max
Checkpoint Remap 0 pages, 0 mapped back. 80 s atomic time.
    DB master 4410112 total 2109374 free 0 remap 0 mapped back
   temp  256 total 251 free

Lock Status: 0 deadlocks of which 0 2r1w, 382 waits,
   Currently 1 threads running 0 threads waiting 0 threads in vdb.
Pending:

Client 1111:8:  Account: dba, 5781 bytes in, 54279 bytes out, 1 stmts.
PID: 0, OS: .NET, Application: .NET Application, IP#: ******
Transaction status: PENDING, 0 threads.
Locks:

Client 1111:9:  Account: dba, 221 bytes in, 256 bytes out, 1 stmts.
PID: 6464, OS: win32, Application: ISQL, IP#: 127.0.0.1
Transaction status: PENDING, 1 threads.
Locks:

Client 1111:6:  Account: dba, 4820 bytes in, 43978 bytes out, 1 stmts.
PID: 0, OS: .NET, Application: .NET Application, IP#: ******
Transaction status: PENDING, 0 threads.
Locks:

Running Statements:
 Time (msec) Text
         189 status()

Hash indexes

47 Rows. -- 234 msec.
SQL>

**Comment by HughWilliams** [comment]
@drassokhin: 

What version of  DotNet or ADO.Net does you client app run on ?

Which Virtuoso ADO.Net Provider is being used are you using the open source version (OpenLink.Data.Virtuoso.dll) which is based on .Net 2 although should be compatible with later versions, from the source archive or the commercial ADO.Net Provider available in the Virtuoso Windows client at http://opldownload.s3.amazonaws.com/uda/virtuoso/7.2/x86_64-generic-win-64/wavcz2zz.msi , which has ADO.Net Providers (virrado3.dll and virtado4.dll) built for  .Net 3.x and .Net 4.x ?

As I think it might be worth trying the commercial providers, if as I would expect you are using a later version of .Net in case they is some subtle incompatibility causing this hang ...

**Comment by drassokhin** [comment]
Hugh, I extracted virtado4.dll from x86_64-generic-win-64/wavcz2zz.msi, changed the OpenLink.Data.Virtuoso.dll reference to virtado4.dll and rebuilt the code. Unfortunately, it did not make any difference. The the code that reads query results still gets stuck randomly at System.dll!System.Net.Sockets.Socket.Poll. As I mentioned before, it's timing-dependent. The number of iterations before Read() hangs changes if I place Thread.Sleep(some_time_interval) inside the recordset reading loop.

BTW, I think it would be a good idea to deploy the ADO.NET drivers as nuget packages the same way they publish the open source OpenLink.Data.Virtuoso.dll. Ideally, if the commercial driver contains any features not present in the open source one, merge those features into the open source package and release it via nuget. If the ADO.NET driver has any "commercial" features that require separate licensing, extract them into a separate commercial package. Having multiple Virtuoso ADO.NET drivers with overlapping functionality but different names creates confusion.

Installing Virtuoso ADO.NET providers system-wide from the msi package and registering them in the Global Assembly Cache is a really bad idea that creates problems for software developers using these drivers. If a developer runs the msi, adds a reference to virtado4.dll in his project and then builds it, the binaries run fine on the developer's computer. However, when the software gets deployed to users' computers, which do not have virtado4.dll in the Global Assembly Cache, it fails at runtime if the developer forgets to extract and include virtado4.dll with his software package. In my case, the software in which I use Virtuoso ADO.NET drivers is a plugin for a large data analysis/visualization application. The plugin does not come with a separate install, but is deployed by simply copying its binaries to a designated plugin folder. It can't register anything in the GAC, modify environment variables or system registry and has to be self-contained with all dependencies deployed to the folder containing the plugin's binaries.

**Comment by drassokhin** [comment]
We still experience those random hangs in PollRead method. Sometimes they occur during the execution of VirtuosoDataReader.Read, sometimes they happen in the ManagedCommand.Dispose() method. I have actually switched my project from the pre-built virtado4.dll to OpenLink.Data.Virtuoso.dll built from the latest source in the dev. branch, but it did not fix the error. The same queries that cause the PollRead to wait indefinitely cause no problems when they are executed via either the isql command line or the SPARQL tab in Virtuoso Conductor. I also noticed that SPARQL queries having the LIMIT nn clause cause this behavior much more often compared to the same queries without the LIMIT.

After a process stuck in PollRead is killed, status() command on the server side still reports an active client session from that process, for example:
Client 1111:3:  Account: xxx, 153940 bytes in, 517298 bytes out, 4 stmts.
PID: 0, OS: .NET, Application: .NET Application, IP#: nnn.nn.nnn.nn
Transaction status: PENDING, 0 threads.
Locks:

**Comment by HughWilliams** [comment]
@drassokhin: Hadn't heard from you regarding this issue and thus assumed it might have been solved with the Cancel method hang fix we provided sometime ago ...

There is an updated ADO.Net 4 Provider (virtado4.dll) available from:

     ftp://download.openlinksw.com/support/virtuoso/virtado4.zip

specifically with a datetime fix, but also other fixes that have been made over the past week. Thus I would suggest testing with this updated provider to see if the problem still occurs.

If the problem still persists if you could provide a LinqPad test app to demonstrate this against he Open Phacts of other public datasets, we could attempt to recreate in-house ...

Can you also provide details on the version of Virtuoso being used (virtuoso-t - ?) , as copy of your Virtuoso INI and log files, and any example queries that cause the hang ...

**Comment by drassokhin** [comment]
@HughWilliams The version of Virtuoso server we are using is "OpenLink Virtuoso  Server
Version 07.20.3217-threads for Win64 as of Apr 25 2016". As to the ADO.NET provider, I compiled it myself from the latest source (from the develop/7 branch as of about 3 weeks ago; the branch history in github does not show any changes to the ADO.NET code made after I synched my copy with the repository). I can try to use the pre-built version, but I don't think it will make any difference, because it has been built from the same code.

This problem is really a pain in the neck, since it is very hard to reproduce. To make things worse, when the driver is stuck in the PollRead method, the Cancel call has no effect, so our users have to quit and restart the application when this happens. I will try to find conditions under which this bug can be reproduced reliably and will let you know. Perhaps, we can even set up a remote desktop session so I can show the developers where exactly the code hangs.

**Comment by HughWilliams** [comment]
@drassokhin:  All the latest ADO.Net Provider fixes would be in the git develop/7 branch ... it still would be worth trying with the pre-built Provider in case it is a build issue ...

If a test case for recreating can be provided that would be ideal ...

**Comment by smalinin** [comment]
Did you call Cancel method from a separate thread ?
Did you change  setting `SQL_PREFETCH_ROW = 100` in virtuoso.ini ?

**Comment by smalinin** [comment]
The Cancel could not have any effect, if the problem isn't related with with too long query execution or with preparing of next resultSet chunk for client. I think. 
Also, did you try to use more fresh build of Virtuoso server? The problem may be related with server side issues also.

**Comment by drassokhin** [comment]
@smalinin No, I have just opened the ini file and confirmed that the SQL_PREFETCH_ROW is set to 100, which I think is the default value.

Yes, of course, Cancel in our application is always called from a separate thread. Not to mention it would be physically impossible to call it from the thread that is stuck on waiting for the PollRead method to return.

As to using a fresh build of Virtuoso server, it would be quite time-consuming for us to try that. We currently run v.7.2.4 pre-built Windows package released on April 25, 2016. We have successfully built Liunx binaries from the Virtuoso Open Source source code in the past. However, we run our production system on a Windows-based host for some technical reasons. Building Windows binaries from source would be quite a hassle, not to mention that the two instances we currently use are the production and staging, and both are in active use, so replacing the pre-built binaries with the latest binaries, which have not been officially tested and released yet, would be a really bad idea.

As I mentioned before, the same queries that cause the ADO.NET driver to hang run just fine and retrieve complete query results w/o any problems when executed via the isql command-line tool or Virtuoso Composer SPARQL tab, so there's something peculiar in how the ADO.NET driver communicates with the server (timing?) that causes this behavior.

**Comment by HughWilliams** [comment]
@drassokhin: You once provide a "VirtuosoNativeDriverTest.linq" test program for a hang querying the Virtuoso server in a direct email to me, is this a test program for recreating this issue or was that another hang condition ?  This was provided around the time of the Cancel problem reported and is what I referred to above when I indicating I thought this problem had already been resolved ...

**Comment by smalinin** [comment]
I have updated the Fetch implementation in ADO.NET provider, it will resolve the problem.

**Comment by smalinin** [comment]
The patch has been commited to Git.

**Comment by HughWilliams** [comment]
@drassokhin: I have verified the fix from @smalinin I have  with the "VirtuosoNativeDriverTest.linq" sample .Net application you provided previously, with the app hanging after about 100K rows with the old Provider, whereas with the patched Provider I can download 500K+ triples without problems and is significantly faster ...

As @smalinin indicates the patch has been committed to the Virtuoso develop/7 git repo, thus you can build from there or the OpenLink.Data.Virtuoso.dll Provider I tested is available from:

   ftp://download.openlinksw.com/support/virtuoso/OpenLink.Data.Virtuoso.dll.zip

**Comment by drassokhin** [comment]
I built the .net Virtuoso driver from the latest source code and re-built our app with the new version of the driver. I tested the latest version by running a bunch of queries, and the query results were retrieved w/o the app hanging, as it used to happen before with the previous version of the .net driver. So far so good. I have released the new version of our app to the users. If they still keep complaining about the app hanging randomly on executing queries, I will let you know.
